src/libcore/cell.rs

```diff
@@ -1290,6 +1298,38 @@ impl<'b, T: ?Sized> RefMut<'b, T> {
         let borrow = orig.borrow.clone();
         (RefMut { value: a, borrow }, RefMut { value: b, borrow: orig.borrow })
     }
+
+    /// Downgrade a `RefMut` to a `Ref`.
+    ///
+    /// The `RefCell` is already mutably borrowed, so this cannot fail.
+    ///
+    /// This is an associated function that needs to be used as
+    /// `RefMut::downgrade(...)`. A method would interfere with methods of the
+    /// same name on the contents of a `RefCell` used through `Deref`.
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(refcell_downgrade)]
+    /// use std::cell::{RefCell, RefMut};
+    ///
+    /// let cell = RefCell::new(vec![1, 2]);
+    /// let mut borrow_mut = cell.borrow_mut();
+    /// borrow_mut.push(3);
+    /// let borrow1 = RefMut::downgrade(borrow_mut);
+    /// let borrow2 = cell.borrow();
+    /// assert_eq!(*borrow1, *borrow2);
+    /// ```
+    #[unstable(feature = "refcell_downgrade", issue = "0")] // FIXME issue number
+    #[inline]
+    pub fn downgrade(orig: RefMut<'b, T>) -> Ref<'b, T> {
+        let borrow = orig.borrow.borrow;
+        mem::drop(orig.borrow);
+
+        Ref {
+            value: orig.value,
+            borrow: BorrowRef::new(borrow).expect("inconsistent RefCell state"),
```

Comment:
This isn't compatible with `map_split`, is it?

Reply:
Right, it'll panic if you try to downgrade after a `RefMut::map_split()` with both references still alive.
Ideally you'd be able to mutably borrow one half and immutably borrow the other like with regular references, but I don't think that's possible with runtime-checked borrows as we can't know in advance how many borrow counters we need to allocate in the `RefCell`.

Still, I can think of a few ways to ameliorate the issue:
- Add new APIs for downgrading multiple references at once
- Add downgrading versions of `map_split` (and `map` for consistency)
- Simply state it as a limitation of the API in the docs

I think the second option is the one I find the most attractive, since it also fixes the related problem of being unable to map with a `(&mut T) -> &U` function.

Reply:
> unable to map with a `(&mut T) -> &U` function

Hm, I think this is unsound (assuming that your downgrading map returns a shared `Ref`, not some special kind of ref) in the presence of functions like [Cell::from_mut](https://doc.rust-lang.org/nightly/std/cell/struct.Cell.html#method.from_mut): you could `.borrow_mut()` to get a `RefMut<T>`, `map_downgrade(Cell::from_mut)` to get a `Ref<Cell<T>>`, then `.borrow()` again to get a `Ref<T>` - but then you could write via the `Ref<Cell<T>>`.

In general `RefMut` -> `Ref` conversion, in the presence of `map`, seems like it could interact badly with other unsafe abstractions...

Reply:
That's a good point, this is trickier than I thought. Reading through the PRs and issues for `Cell::from_mut` it looks like its soundness was proven pretty conclusively, so I'm wondering if `map_downgrade` is unsound in general or if all kinds of downgrading would be.

Reply:
It's not just `Cell::from_mut`. I think `Mutex` is also incompatible with `map_downgrade`, at least if it allows calling `borrow` again after the downgrade: Starting with `RefCell<Mutex<T>>`, you do `.borrow_mut().map_downgrade(|m| &*m.get_mut())`, so you end up with a `Ref<T>` but the mutex is *not* marked as locked.  Then you do `.borrow().lock().unwrap()`, and you can write into the same memory. (This is a variant of [this counterexample for another proposal](https://internals.rust-lang.org/t/blog-post-nested-method-calls-via-two-phase-borrowing/4886/33?u=nikomatsakis).)